Properly handle `CSGShape` parent and visibility updates, plus some refactoring

hoontee opened the pull request:
- `CSGShapes` no longer lose collision when the root shape is made invisible. Fixes #43251.
- Prevents unnecessary `CSGShape` updates when changing the parent's visibility or hierarchy. Fixes godotengine#40931.
- Allows `_update_shape` to be used on invisible shapes. Fixes godotengine#41093.
- Closes #43381.
- Refactor: Renames `parent` to `parent_shape`.
- Refactor: Cleaned up `CSGShape3D::_notification`.

##
Test projects:
[40814 Test Suite - Visibility.zip](https://github.com/godotengine/godot/files/5044638/40814.Test.Suite.-.Visibility.zip)
[40814 Test Suite - Scene Tree.zip](https://github.com/godotengine/godot/files/5044637/40814.Test.Suite.-.Scene.Tree.zip)
[40814 Test Suite - Collision.zip](https://github.com/godotengine/godot/files/5044952/40814.Test.Suite.-.Collision.zip)

madmiraal reviewed (commented):
This is not the solution to #40782, and it may cause other problems.

First, the problem is not that the `CSGShape` is being updated unnecessarily, the problem is that when the `CSGShape` is updated it's not being rendered properly. By avoiding updating the `CSGShape` you're just delaying the issue until the `CSGShape` does need to be updated.

Second, just because the parent or visibility of the child hasn't changed, doesn't necessarily mean that the `CSGShape` doesn't need to be updated. I don't think the efficiency gain of avoiding an update outweighs the risk of missing an update.

hoontee commented:
Could you name a case where changing the root `CSGShape`'s parent or visibility would require an update? I had done some testing with these changes and found no ill effects.

I do agree that it isn't a proper fix for the linked issue, however this does fix another issue I was personally experiencing where the game would stutter after toggling visibility on a large amount of `CSGShapes` (for culling reasons.)

madmiraal commented:
As an example, take the minimal reproduction project provided in #40782 and make the `CSGContainer` node start invisible, with this PR the `CSGBox` is never displayed.

hoontee commented:
You're totally right. I had encountered that during testing and thought I fixed it for some reason.
I'll go ahead and do more testing at runtime just to confirm that there aren't any other bugs I'm missing.

hoontee commented on modules/csg/csg_shape.cpp:
```diff
@@ -160,7 +160,7 @@ CSGBrush *CSGShape3D::_get_brush() {
 				continue;
 			}
-			if (!child->is_visible_in_tree()) {
+			if (!child->is_visible()) {
```
Fix for https://github.com/godotengine/godot/pull/40814#issuecomment-666486128.

hoontee commented:
Can confirm that baked lighting still breaks / un-breaks when toggling visibility of child shapes.

hoontee commented:
Including my changes, godotengine#40782 now only seems to occur when toggling visibility of a single child shape inside a `CSGCombiner`.
Previously it would occur when doing that and when toggling visibility of the `CSGCombiner`.
I'll continue attempting fixes before marking the PR as ready for review again.

hoontee commented on modules/csg/csg_shape.cpp:
```diff
@@ -646,7 +652,7 @@ CSGShape3D::~CSGShape3D() {
 
 CSGBrush *CSGCombiner3D::_build_brush() {
-	return nullptr; //does not build anything
+	return memnew(CSGBrush); //does not build anything
```
Fix for #40782.

hoontee commented:
Also, it may be important to note that #40880 (9be7b30f3536057b22fd70d5367268749c1a7ea9) mentions `CSGShapes` becoming `dirty` after exiting the scene tree, which was intentionally removed in this PR as part of its goal.

However, having done further testing with both this PR and the commit, they tackle separate issues and thereby don't affect each other.

madmiraal commented:
> @madmiraal Found a true fix for #40782 -> #40814 (comment)
I can split #40814 (comment) into its own PR if you'd prefer. Please let me know if you have any further concerns.

Well done! Yes, please split this fix into a separate PR; so it can be merged and cherry-picked for 3.2.

hoontee commented:
@madmiraal OK, it's been posted! #40919
I hope you may still consider merging this PR in the future as it does still fix issues I was having with stuttering when toggling portions of the level.

hoontee commented:
Updated to fix an issue with collisions not working.

clayjohn commented:
@hoontee unrelated to this PR, but just for your information, you don't need to close the PR before force pushing. In fact, it's better to just leave the PR open. It will also result in fewer notifications for those of us watching the repo. :)

hoontee commented:
@clayjohn It keeps doing that automatically due to there being no commits. I'm clearly not using git correctly, haha.
Since it notifies everyone I'll try figuring out how to improve the way I make changes on a separate repo.
```
git push --force origin 227494b:master
git reset --hard origin/master
git add "modules/csg/csg_shape.cpp"
git add "modules/csg/csg_shape.h"
git commit
git push
```

hoontee commented:
Updated to fix an issue with shapes not updating when modified outside the scene tree.
Had to rebase, but I didn't do it properly so sorry for the spam :(

Xrayez commented:
Linking #25783 because related, was long time ago...

hoontee commented:
This PR probably should have been a draft prior to today, however I feel it's now in a truly conpleted state.
I would appreciate it if someone would test this PR further to see if there are any other bugs left unfound. Here's what I tested:
- Toggling visibility / changing parent of / transforming / modifying shapes with children, child shapes, and shapes with no children, both inside and outside the scene tree
- Loading scene with invisible shapes with children, child shapes, and shapes with no children and later setting to visible

madmiraal reviewed (changes requested):
I'm still concerned that there are missed test cases. Please attach a copy of your test project(s) so that we can confirm what has been tested. Also, as specified below, please add comments to explain exactly what is being tested for when `_mark_dirty()` should not be called. This will make it easier to understand and troubleshoot; especially if bugs are reported later.

Otherwise, please see below for my specific concerns regarding the code.

hoontee commented on modules/csg/csg_shape.cpp:
```diff
@@ -503,7 +504,13 @@ void CSGShape3D::_notification(int p_what) {
 
-		_make_dirty();
+		if (!_get_brush() || (parentn && last_parent != Object::cast_to<CSGShape3D>(parentn))) {
+			_make_dirty();
```
`_get_brush` returns the current `brush` if not `dirty`. The optimization comes with preventing `_make_dirty` calls.

However, this check was added very early on so it may not still be necessary. I'll review this first thing tomorrow.

hoontee replied:
This check was indeed unnecessary, so it's been removed.

hoontee commented on modules/csg/csg_shape.h:
```diff
@@ -52,12 +52,14 @@ class CSGShape3D : public GeometryInstance3D {
 	Operation operation;
 	CSGShape3D *parent;
+	CSGShape3D *last_parent;
 
```
It's a `CSGShape` for type checking purposes. We don't want to update the shape if its parent switches from a non-shape to another non-shape.

hoontee commented on modules/csg/csg_shape.cpp:
```diff
@@ -503,7 +504,13 @@ void CSGShape3D::_notification(int p_what) {
+			_make_dirty();
+		} else {
+			_update_collision_faces();
+		}
```
Is this change suitable? https://github.com/hoontee/godot/blob/12e2870fbac00bc0d8d1a253c81213e161798b04/modules/csg/csg_shape.cpp#L431-L432

hoontee commented:
`CSGShape3D::_notification` has been refactored for simplicity. [(view the change)](https://github.com/hoontee/godot/blob/3de8d58de2615d4de0b0390dbfdcafd3bb4e96b2/modules/csg/csg_shape.cpp#L485-L554)

hoontee commented:
Added two test projects:
[40814 Test Suite - Visibility.zip](https://github.com/godotengine/godot/files/5044638/40814.Test.Suite.-.Visibility.zip)
[40814 Test Suite - Scene Tree.zip](https://github.com/godotengine/godot/files/5044637/40814.Test.Suite.-.Scene.Tree.zip)

hoontee commented:
Updated to explain why `_update_shape` gets deferred. [(view)](https://github.com/hoontee/godot/blob/25a63df2f40fa776a13db2ae8aee75d16fad2e0c/modules/csg/csg_shape.cpp#L132)
This would also explain why it's exposed since `call_deferred` doesn't work on private methods.

hoontee commented:
Updated to re-add a small snippet of code and added an explanation for why it's needed. [(view)](https://github.com/hoontee/godot/blob/45f3ebced62913c76c721837a50d4edeb8eebbe4/modules/csg/csg_shape.cpp#L492-L496)

hoontee commented:
Ready for review.

hoontee commented:
Rebased following 920f8c86ec7fe6172365b89a7c290af956e8e8e3 and 3044b8f15af91aac4177c59bb5455dcd33a1a6c3.
Still don't know how to rebase without getting the PR automatically closed. :/

akien-mga commented:
> ```
> git push --force origin 227494b:master
> git reset --hard origin/master
> git add "modules/csg/csg_shape.cpp"
> git add "modules/csg/csg_shape.h"
> git commit
> git push
> ```

@hoontee The PR will be closed automatically if you force push an empty changeset with regard to the `master` branch, as there isn't anything to merge. You should only force push once you are done with your local changes, not before.

See https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

hoontee commented:
As of 3.2.4 beta 1, CSG collision no longer works when the shape is invisible. After some testing I can confirm that this PR fixes that issue as well.

aaronfranke commented on modules/csg/csg_shape.cpp:
```diff
@@ -127,15 +127,12 @@ float CSGShape3D::get_snap() const {
-void CSGShape3D::_make_dirty() {
-	if (!is_inside_tree()) {
-		return;
+void CSGShape3D::_make_dirty(bool parent_removing) {
+	if ((parent_removing || is_root_shape()) && !dirty) {
+		call_deferred("_update_shape"); // Must be deferred; otherwise, is_root_shape() will use the previous parent
```
Parameters start with `p_` in the C++ codebase.

```suggestion
void CSGShape3D::_make_dirty(bool p_parent_removing) {
	if ((p_parent_removing || is_root_shape()) && !dirty) {
		call_deferred("_update_shape"); // Must be deferred; otherwise, is_root_shape() will use the previous parent.
```

aaronfranke commented on modules/csg/csg_shape.cpp:
```diff
@@ -482,60 +484,67 @@ Vector<Face3> CSGShape3D::get_faces(uint32_t p_usage_flags) const {
-				set_base(RID());
-				root_mesh.unref();
+	switch (p_what) {
+		case (NOTIFICATION_PARENTED): {
```
I think the parenthesis on the case are not necessary.

akien-mga commented:
Superseded by #43322.

hoontee commented:
Reopened as an alternative to both #43322 and #43381.

hoontee committed:
Properly handle CSGShape parent and visibility updates, plus some refactoring

hoontee commented:
Rebased following f7209b459b4faaae9d93bfb6ac5346eb41787f92.

I tried to follow [the pr workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) but I ended up pushing another empty changeset.

hoontee commented:
Needs the `bug` and `regression` labels.
The `needs testing` label may also need to be removed - I've also tested this thoroughly but I'm unsure if others need to test it as well.

hoontee commented:
Rebased following #52647.

hoontee commented:
Tested the fixes again and they're still working. I just wish someone could review this...

hoontee commented:
The fix still works. Now that the 4.0 alpha is out, 1.5 years after submitting this PR, perhaps it could be reviewed?

hoontee commented:
Rebased following 6553f5c242865f9aadbe2fdab6db6876c44ac472.

akien-mga reviewed (approved):
CSG is an area where we currently lack dedicated maintainers, and so I've been trying to encourage other contributors to review and test this PR but without much luck...

After discussion with @reduz we agreed to merge it as is, and see how it fares in production. Sorry for the huge delay and lack of responsiveness.

akien-mga commented:
Thanks!
